File: backend/layer2bridge/layer2interface.py
import requests
import json
import random
import string
from typing import List, Dict, Any
from dataclasses import dataclass
import DatabaseInterface
import AuditDatabaseInterface
from openl2_messaging import (
    signDepositMessage,
    signWithdrawalBroadcastedMessage,
    signWithdrawalConfirmedMessage,
    signLayer1AuditReportMessage
)
from OnboardingLogger import OnboardingLogger


DEFAULT_LAYER2_URL = 'https://testnet.instachain.io/'

ERROR_SUCCESS = 0
ERROR_UNKNOWN = 1
ERROR_CANNOT_VERIFY_SIGNATURE = 10
ERROR_CANNOT_DUPLICATE_TRANSACTION = 11
ERROR_INSUFFICIENT_FUNDS = 12
ERROR_TRANSACTION_ID_NOT_FOUND = 13
ERROR_ONBOARDING_PUBKEY_MISMATCH = 14
ERROR_CANNOT_CANCEL_WITHDRAWAL_MULTIPLE_TIMES = 15
ERROR_DEPOSIT_ADDRESS_NOT_FOUND = 16
ERROR_DUPLICATE_TRANSACTION_ID = 17
ERROR_COULD_NOT_FIND_WITHDRAWAL_REQUEST = 18
ERROR_DATABASE_TRANSACTIONAL_ERROR = 19

def SuccessOrDuplicateErrorCode(error: int):
    return error in (ERROR_SUCCESS, ERROR_CANNOT_DUPLICATE_TRANSACTION, ERROR_DUPLICATE_TRANSACTION_ID)

class Layer2Interface:
    layer2_node_url: str
    onboarding_signing_private_key: str

    @dataclass
    class WithdrawalBroadcastedTransaction:
        layer1_transaction_id: str
        layer1_transaction_vout: int
        layer1_address: str
        amount: int
        layer2_withdrawal_id: str
        signature: str = ""

    def __init__(self, layer2_node_url: str, onboarding_signing_private_key: str):
        self.layer2_node_url = layer2_node_url or DEFAULT_LAYER2_URL
        self.onboarding_signing_private_key = onboarding_signing_private_key

    header = {'user-agent': 'requests/0.0.1'}
    def getWithdrawalRequests(self, lastwithdrawalTimestamp: int):
        url = self.layer2_node_url + 'getWithdrawalRequests'
        data = {'latest_timestamp': lastwithdrawalTimestamp}
        r = requests.get(url, params=data, headers=self.header)
        OnboardingLogger(r.text)
        return r.text

    def ackWithdrawalRequests(self, layer2_withdrawal_ids: List[str]):
        url = self.layer2_node_url + 'ackWithdrawalRequests'
        data = {'layer2_withdrawal_ids': layer2_withdrawal_ids}
        r = requests.post(url, params=data, headers=self.header)
        OnboardingLogger(r.text)
        return r.text

    def confirmDeposit(self, nonce: str, layer1_transaction_id: str, amount: int, layer1_address: str, signature: str):
        url = self.layer2_node_url + 'depositFunds'
        data = {'nonce': nonce,
                'layer1_transaction_id': layer1_transaction_id,
                'amount': amount,
                'layer1_address': layer1_address,
                'signature': signature}
        r = requests.post(url, params=data, headers=self.header)
        OnboardingLogger('/confirmDeposit ' + layer1_transaction_id)
        OnboardingLogger(r.text)
        return r.text

    def confirmDepositMulti(self, depositTransactions: List[DatabaseInterface.ConfirmedTransaction]):
        url = self.layer2_node_url + 'depositFunds'
        transactions = []
        for depositTransaction in depositTransactions:
            transaction = {'layer1_transaction_id': depositTransaction.transaction_id,
                    'layer1_transaction_vout': depositTransaction.transaction_vout,
                    'amount': depositTransaction.amount,
                    'layer1_address': depositTransaction.address,
                    'nonce': depositTransaction.nonce,
                    'signature': depositTransaction.signature}
            transactions.append(transaction)
        jsonData = {"transactions":transactions}
        r = requests.post(url, json=jsonData, headers=self.header)
        OnboardingLogger("confirmDepositMulti: " + str(r.text))
        return r.text

    def broadcastWithdrawalMulti(self, withdrawalBroadcastedTransactions: List[WithdrawalBroadcastedTransaction]):
        url = self.layer2_node_url + 'withdrawalBroadcasted'
        transactions = []
        for withdrawalBroadcastedTransaction in withdrawalBroadcastedTransactions:
            transaction = {"layer1_transaction_id": withdrawalBroadcastedTransaction.layer1_transaction_id,
                "layer1_transaction_vout": withdrawalBroadcastedTransaction.layer1_transaction_vout,
                "layer1_address": withdrawalBroadcastedTransaction.layer1_address,
                "amount": withdrawalBroadcastedTransaction.amount,
                "layer2_withdrawal_id": withdrawalBroadcastedTransaction.layer2_withdrawal_id,
                "signature": withdrawalBroadcastedTransaction.signature}

            transactions.append(transaction)
        jsonData = {"transactions":transactions}
        OnboardingLogger('broadcastWithdrawalMulti: ' + json.dumps(jsonData))
        r = requests.post(url, json=jsonData, headers=self.header)
        OnboardingLogger(r.text)
        return r.text

    def broadcastWithdrawal(self, layer1_transaction_id: str, layer1_transaction_vout: int, layer1_address: str, amount: int, layer2_withdrawal_id: str, signature: str):
        url = self.layer2_node_url + 'withdrawalBroadcasted'
        data = {'layer1_transaction_id': layer1_transaction_id,
                'layer1_transaction_vout': layer1_transaction_vout,
                'layer1_address': layer1_address,
                'amount': amount,
                'layer2_withdrawal_id': layer2_withdrawal_id,
                'signature': signature}
        r = requests.post(url, params=data, headers=self.header)
        OnboardingLogger(r.text)
        return r.text

    def confirmWithdrawal(self, layer1_transaction_id: str, layer1_transaction_vout: int, layer1_address: str, amount: int, signature: str):
        url = self.layer2_node_url + 'withdrawalConfirmed'
        data = {'layer1_transaction_id': layer1_transaction_id,
                'layer1_transaction_vout': layer1_transaction_vout,
                'layer1_address': layer1_address,
                'amount': amount,
                'signature': signature}
        r = requests.post(url, params=data, headers=self.header)
        OnboardingLogger(r.text)
        return r.text

    def confirmWithdrawalMulti(self, confirmedWithdrawals: List[DatabaseInterface.ConfirmedTransaction]):
        url = self.layer2_node_url + 'withdrawalConfirmed'
        transactions = []
        for confirmedWithdrawal in confirmedWithdrawals:
            transaction = {'layer1_transaction_id': confirmedWithdrawal.transaction_id,
                'layer1_transaction_vout': confirmedWithdrawal.transaction_vout,
                'layer1_address': confirmedWithdrawal.address,
                'amount': confirmedWithdrawal.amount,
                'signature': confirmedWithdrawal.signature}
            transactions.append(transaction)
        jsonData = {"transactions":transactions}
        r = requests.post(url, json=jsonData, headers=self.header)
        OnboardingLogger(r.text)
        return r.text
    
    def postLayer1AuditReport(self, blockheight: int, balance: int, layer1AddressBalances: List[AuditDatabaseInterface.AuditLayer1Address]):
        url = self.layer2_node_url + 'postLayer1AuditReport'
        layer1AddressBalancesJson = [layer1AddressBalance.to_dict() for layer1AddressBalance in layer1AddressBalances]
        signature = signLayer1AuditReportMessage(self.onboarding_signing_private_key, blockheight, balance)
        jsonData = {'block_height': blockheight,
                'balance': balance,
                'layer1_address_balances': layer1AddressBalancesJson,
                'signature': signature}
        r = requests.post(url, json=jsonData, headers=self.header)
        OnboardingLogger(r.text)
        return r.text


    def sendConfirmDeposit(self, transactions: List[DatabaseInterface.ConfirmedTransaction]):
        for transaction in transactions:
            transaction.nonce = ''.join(random.choice(string.ascii_letters) for i in range(16))
            transaction.signature = signDepositMessage(self.onboarding_signing_private_key, transaction.transaction_id, transaction.transaction_vout, transaction.address, transaction.amount, transaction.nonce)
        return self.confirmDepositMulti(transactions)

    def sendWithdrawalBroadcasted(self, withdrawalBroadcastedTransactions: List[WithdrawalBroadcastedTransaction]):
        for withdrawalBroadcastedTransaction in withdrawalBroadcastedTransactions:
            withdrawalBroadcastedTransaction.signature = signWithdrawalBroadcastedMessage(self.onboarding_signing_private_key, withdrawalBroadcastedTransaction.layer1_transaction_id, withdrawalBroadcastedTransaction.layer1_transaction_vout, withdrawalBroadcastedTransaction.layer1_address, withdrawalBroadcastedTransaction.amount, withdrawalBroadcastedTransaction.layer2_withdrawal_id)
        return self.broadcastWithdrawalMulti(withdrawalBroadcastedTransactions)

    def sendConfirmWithdrawal(self, confirmedWithdrawals: List[DatabaseInterface.ConfirmedTransaction]):
        for transaction in confirmedWithdrawals:
            transaction.signature = signWithdrawalConfirmedMessage(self.onboarding_signing_private_key, transaction.transaction_id, transaction.transaction_vout, transaction.address, transaction.amount)
        return self.confirmWithdrawalMulti(confirmedWithdrawals)

File: backend/layer2bridge/src/layer2-interface.ts
import {
	createLayer2LedgerClient,
	type DepositConfirmedRequest,
	type DepositConfirmedResponse,
	type DepositsConfirmed,
	ErrorCodes,
	type GetWithdrawalRequestsRequest,
	type GetWithdrawalRequestsResponse,
	type Layer1BroadcastedWithdrawalTransaction,
	type Layer1WithdrawalConfirmedTransaction,
	type Layer2LedgerClient,
	unwrapLayer2LedgerResponse,
	type WithdrawalBroadcastedRequest,
	type WithdrawalBroadcastedResponse,
	type WithdrawalConfirmedRequest,
	type WithdrawalConfirmedResponse,
} from "@openl2/api-layer2ledger";
import {
	buildDepositMessage,
	buildLayer1AuditReportMessage,
	buildWithdrawalBroadcastedMessage,
	buildWithdrawalConfirmedMessage,
} from "@openl2/openl2-messaging";
import { signMessage } from "@openl2/pubkey-utils";

export function successOrDuplicateErrorCode(error: number): boolean {
	return (
		error === ErrorCodes.SUCCESS ||
		error === ErrorCodes.CANNOT_DUPLICATE_TRANSACTION ||
		error === ErrorCodes.DUPLICATE_TRANSACTION_ID
	);
}

export interface Layer1TransactionConfirmationInput {
	transactionId: string;
	transactionVout: number;
	address: string;
	amount: number;
}

export type WithdrawalBroadcastInput = Omit<
	Layer1BroadcastedWithdrawalTransaction,
	"signature"
>;

export interface Layer1AddressBalanceInput {
	layer1_address: string;
	balance: number;
}

export interface PostLayer1AuditReportResponse {
	error_code: number;
}

const AUDIT_ROUTER_PREFIX = "/audit";
const POST_LAYER1_AUDIT_REPORT_ROUTE = "/post_layer1_audit_report";

export class Layer2Interface {
	private readonly client: Layer2LedgerClient;
	private readonly layer2NodeUrl: string;

	constructor(
		layer2NodeUrl: string,
		private readonly onboardingSigningPrivateKey: string,
		private readonly nodeId: string,
	) {
		this.layer2NodeUrl = layer2NodeUrl.endsWith("/")
			? layer2NodeUrl.slice(0, -1)
			: layer2NodeUrl;
		this.client = createLayer2LedgerClient(this.layer2NodeUrl);
	}

	async getWithdrawalRequests(
		lastWithdrawalTimestamp: number,
	): Promise<GetWithdrawalRequestsResponse> {
		const body: GetWithdrawalRequestsRequest = {
			latest_timestamp: lastWithdrawalTimestamp,
		};
		const result =
			await this.client.withdrawal.get_withdrawal_requests.post(body);
		return unwrapLayer2LedgerResponse(result);
	}

	async confirmDepositMulti(
		transactions: DepositsConfirmed[],
	): Promise<DepositConfirmedResponse> {
		const body: DepositConfirmedRequest = { transactions };
		const result = await this.client.deposit.deposit_confirmed.post(body);
		return unwrapLayer2LedgerResponse(result);
	}

	async broadcastWithdrawalMulti(
		transactions: Layer1BroadcastedWithdrawalTransaction[],
	): Promise<WithdrawalBroadcastedResponse> {
		const body: WithdrawalBroadcastedRequest = { transactions };
		const result =
			await this.client.withdrawal.withdrawal_broadcasted.post(body);
		return unwrapLayer2LedgerResponse(result);
	}

	async confirmWithdrawalMulti(
		transactions: Layer1WithdrawalConfirmedTransaction[],
	): Promise<WithdrawalConfirmedResponse> {
		const body: WithdrawalConfirmedRequest = { transactions };
		const result = await this.client.withdrawal.withdrawal_confirmed.post(body);
		return unwrapLayer2LedgerResponse(result);
	}

	async sendConfirmDeposit(
		transactions: Layer1TransactionConfirmationInput[],
	): Promise<DepositConfirmedResponse> {
		const depositTxs: DepositsConfirmed[] = [];
		for (const transaction of transactions) {
			const nonce = crypto.randomUUID().replace(/-/g, "").slice(0, 16);
			const message = buildDepositMessage(
				this.nodeId,
				transaction.transactionId,
				transaction.transactionVout,
				transaction.address,
				transaction.amount,
				nonce,
			);
			const signature = await signMessage(
				message,
				this.onboardingSigningPrivateKey,
			);
			depositTxs.push({
				layer1_transaction_id: transaction.transactionId,
				layer1_transaction_vout: transaction.transactionVout,
				amount: transaction.amount,
				layer1_address: transaction.address,
				nonce,
				signature,
			});
		}
		return this.confirmDepositMulti(depositTxs);
	}

	async sendWithdrawalBroadcasted(
		transactions: WithdrawalBroadcastInput[],
	): Promise<WithdrawalBroadcastedResponse> {
		const signed: Layer1BroadcastedWithdrawalTransaction[] = [];
		for (const tx of transactions) {
			const message = buildWithdrawalBroadcastedMessage(
				this.nodeId,
				tx.layer1_transaction_id,
				tx.layer1_transaction_vout,
				tx.layer1_address,
				tx.amount,
				tx.layer2_withdrawal_id,
			);
			signed.push({
				...tx,
				signature: await signMessage(message, this.onboardingSigningPrivateKey),
			});
		}
		return this.broadcastWithdrawalMulti(signed);
	}

	async sendConfirmWithdrawal(
		transactions: Layer1TransactionConfirmationInput[],
	): Promise<WithdrawalConfirmedResponse> {
		const withdrawalTxs: Layer1WithdrawalConfirmedTransaction[] = [];
		for (const transaction of transactions) {
			const message = buildWithdrawalConfirmedMessage(
				this.nodeId,
				transaction.transactionId,
				transaction.transactionVout,
				transaction.address,
				transaction.amount,
			);
			withdrawalTxs.push({
				layer1_transaction_id: transaction.transactionId,
				layer1_transaction_vout: transaction.transactionVout,
				layer1_address: transaction.address,
				amount: transaction.amount,
				signature: await signMessage(message, this.onboardingSigningPrivateKey),
			});
		}
		return this.confirmWithdrawalMulti(withdrawalTxs);
	}

	async postLayer1AuditReport(
		blockHeight: number,
		balance: number,
		layer1AddressBalances: Layer1AddressBalanceInput[],
	): Promise<PostLayer1AuditReportResponse> {
		const message = buildLayer1AuditReportMessage(
			this.nodeId,
			blockHeight,
			balance,
		);
		const signature = await signMessage(
			message,
			this.onboardingSigningPrivateKey,
		);
		const url = `${this.layer2NodeUrl}${AUDIT_ROUTER_PREFIX}${POST_LAYER1_AUDIT_REPORT_ROUTE}`;
		const response = await fetch(url, {
			method: "POST",
			headers: { "Content-Type": "application/json" },
			body: JSON.stringify({
				block_height: blockHeight,
				layer1_address_balances: layer1AddressBalances,
				signature,
			}),
		});
		if (!response.ok) {
			throw new Error(`postLayer1AuditReport failed: HTTP ${response.status}`);
		}
		return (await response.json()) as PostLayer1AuditReportResponse;
	}
}
